sklearn/tree/_criterion.pxd

```diff
@@ -71,6 +57,27 @@ cdef class Criterion:
     ) nogil
     cdef double proxy_impurity_improvement(self) nogil
 
+cdef class Criterion(BaseCriterion):
+    """Abstract interface for supervised impurity criteria."""
+
+    cdef const DOUBLE_t[:, ::1] y
+
+    cdef int init(
+        self,
+        const DOUBLE_t[:, ::1] y,
+        const DOUBLE_t[:] sample_weight,
+        double weighted_n_samples,
+        const SIZE_t[:] sample_indices,
+        SIZE_t start,
+        SIZE_t end
+    ) nogil except -1
+
+    cdef int pointer_reset(
+        self,
+        SIZE_t start,
+        SIZE_t end
+    ) nogil except -1
```

Comment:
The `pointer_reset` function should be defined as an abstract function for `BaseCriterion` because it is a common procedure done no matter what type of criterion function you are computing.

Reply:
Thank you @adam2392! I realized that when I looked at your note you shared with me. I just pushed commits for this. I also made the function not return anything as it was shown on your note, but I carefully think that may not matter much?